builder/dockerfile/internals.go

```diff
@@ -489,9 +489,13 @@ func hostConfigFromOptions(options *types.ImageBuildOptions, isWCOW bool) *conta
 	// For WCOW, the default of 20GB hard-coded in the platform
 	// is too small for builder scenarios where many users are
 	// using RUN statements to install large amounts of data.
-	// Use 127GB as that's the default size of a VHD in Hyper-V.
-	if isWCOW {
-		hc.StorageOpt = make(map[string]string)
+	// If no custom size is set, use 127GB for containers that
+	// are started during build, as that's the default size of
+	// a VHD in Hyper-V.
+	if isWCOW && hc.StorageOpt["size"] == "" {
```

Comment:
How can `hc.StorageOpt` ever be set here? hc is defined 14 lines earlier, and doesn't set the `StorageOpt` member, so won't it always be `nil` here?

Reply:
I think maybe this fix is in the wrong place. The values set here are merged with the daemon's config values later, in `create` in _daemon/create.go_, which is called with the output of this method. See line 190 of _daemon/create.go_. There, the logic is for the StorageOpts created here to override the daemon config, but it should be the other way 'round since the StorageOpts here are defaults.

Reply:
🤦 You are absolutely right. Thinking how I came to this change at the time, as it's a while, but looks like the daemon config is merged in https://github.com/moby/moby/blob/8a4671fb1f89d222bd65cb07ea6a0de1ec7a1a2b/daemon/create.go#L195-L205

🤔

Reply:
Alternatively, move this entire chunk of logic to set a default sandbox size, into _daemon/create.go_ _after_ the merge. It makes now different _now_, but if we ever find a way to get `docker build` to pass storage opts, then we'd want _those_ to win over the daemon.json _and_ the default 127GB sandbox.

Reply:
Perhaps we should hold this change until after we cut the 20.10 branch, and when we start removing the LCOW paths (I'd also personally be ok with increasing the size to a higher value if 127GB is still too low for some scenarios); IIRC, setting a higher value shouldn't affect performance when creating the container (and would only make a difference if someone _actually_ used the given amount of space)

Reply:
The advantage of moving this default-size logic into moby/daemon/create.go is that I think we know if it's LCOW or WCOW by then, because we have `os` set.

Reply:
On the other hand, taking it out of builder/dockerfile means that we can't as-easily reuse the number for the copy mount, as part of fixing #37352.

Reply:
_Easiest_ approach might be to leave _builder/dockerfile/internals.go_ as-is, or with only the `isWCOW` test fixed, and just change daemon/create.go from
```
 		if _, ok := ctr.HostConfig.StorageOpt[opt[0]]; !ok { 
 			ctr.HostConfig.StorageOpt[opt[0]] = opt[1] 
 		} 
```

to just
```
 		ctr.HostConfig.StorageOpt[opt[0]] = opt[1] 
```
, i.e.. `daemon.configStore.GraphOptions` always overrides `ctr.HostConfig.StorageOpt`, and today the only thing that this affects is `size=127GB` from _builder/dockerfile/internals.go_.